src/bsk201910/rollbacks.md

```diff
@@ -0,0 +1,407 @@
+
+_In general, the YAML for the Deployment contains the ReplicaSet plus some additional details._
+
+You can create the ReplicaSet with:
+
+```terminal|command=1|title=bash
+kubectl create -f replicaset.yaml
+```
+
+_There's something else worth noting about the ReplicaSets and Deployments._
+
+When you upgrade your Pods from version 1 to version 2, the Deployment creates a new ReplicaSet and increases the count of replicas while the previous count goes to zero.
+
+**After the rolling update, the previous ReplicaSet is not deleted** — not immediately at least.
+
+Instead, it is kept around with a replicas count of 0.
+
+If you try to execute another rolling update from version 2 to version 3, you might notice that at the end of the upgrade, you have two ReplicaSets with a count of 0.
+
+_Why are the previous ReplicaSets not deleted or garbage collected?_
+
+Imagine that the current version of the container introduces a regression.
+
+You probably don't want to serve unhealthy responses to your users, so you might want to roll back to a previous version of your app.
+
+If you still have an old ReplicaSet, perhaps you could scale the current replicas to zero and increment the previous ReplicaSet count.
+
+In other words, **keeping the previous ReplicaSets around is a convenient mechanism to roll back to a previously working version of your app.**
+
+By default Kubernetes stores the last 10 ReplicaSets and lets you roll back to any of them.
+
+But you can change how many ReplicaSets should be retained by changing the `Deployment.spec.revisionHistoryLimit` in your deployment.
+
+```yaml|highlight=15|title=deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      revisionHistoryLimit: 100
+      containers:
+      - name: app
+        image: learnk8s/hello:1.0.0
+```
+
+_What about the previous ReplicaSets?_
+
+_Could you list all the previous Replicasets that belong to a Deployment?_
+
+You can use the following command to inspect the history of your Deployment:
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+```
+
+And you can rollback to a specific version with:
+
+```terminal|command=1|title=bash
+kubectl rollout undo deployment/app --to-revision=2
+```
+
+_But how does the Deployment know their ReplicaSets?_
+
+_Does it store the order in which ReplicaSets are created?_
+
+The ReplicaSets have random names with id such as `app-6ff88c4474`, so you should expect the Deployment to store a reference to them.
+
+Let's inspect the Deployment with:
+
+```terminal|command=1|title=bash
+kubectl get deployment app -o yaml
+```
+
+Nothing is looking like a list of previous 10 ReplicaSets.
+
+Deployments don't hold a reference to their ReplicaSets.
+
+At least not in the same YAML.
```

Comment:
There is a back reference in the `ReplicaSet` to the `Deployment` it belongs to.  It's a usual compsci trick, to represent an array, we simply have backpointers from the items to the holder of the array...

I don't know if this has to be part of the article, just saying.